### PR TITLE
Read a CodeRabbit finding's title from prose, not quoted code

### PR DESCRIPTION
**Pull request by dlgpdmsly2**
## Summary

`classifyCodeRabbitComment` and `parseCodeRabbitReview` took a CodeRabbit finding's `summary` as the first `**…**` span anywhere in the comment body. The search now skips fenced blocks and HTML comments — the two places where a `**` is markup rather than emphasis — and both readers share one function instead of each carrying a copy of the regex.

Also corrects a comment on `CR_PROSE_END` that asserted something the data contradicts. The constant itself is unchanged.

## Why

Most of a CodeRabbit body is not prose. It carries the web queries it ran, its shell transcripts, a committable suggestion and a `🤖 Prompt for AI Agents` block — and `**` is ordinary shell and regex syntax, so a quoted command contains bolded-looking spans that are not emphasis.

On comment [3651715274](https://github.com/wafflebase/wafflebase/pull/549#discussion_r3651715274) the finding's summary was the string `/node_modules/`, taken from the exclude glob of an `rg` invocation 2129 bytes past where that comment's prose ends. The title CodeRabbit wrote — *"Pin the markdown-it dependency or switch to the public `getRules()` API."* — is in the same body, below the analysis block. Nothing failed; the parser produced a plausible-looking summary and carried on.

Measured over all 2061 CodeRabbit inline comments and all 614 review bodies in this repository: **28 of 1588 inline findings and 1 of 1693 review-body findings** had a title mined out of quoted code. Several are not short globs — one summary is 1.4 KB of shell transcript, another an embedded Python script. Those strings reach `summaryTokens`, `findingKey` and the verifier prompt as a finding's one-line title.

Narrowing the search to stop at the first structured block was the obvious alternative and is strictly worse: a narrower search can only return less, so it cannot correct a wrong title, only delete one. Measured over the same 1588 findings it changed 200 and **every one went from a real title to `""`**, because 12.6% of CodeRabbit's comments open with an `🧩 Analysis chain` block and state their title after it — which is the claim `CR_PROSE_END`'s comment got wrong, and which this PR corrects in place.

`parseCodeRabbitReview` had its own copy of the regex, so it could have regressed independently. One detail there is load-bearing and commented as such: the title is read from the raw span rather than the de-locatored copy `detail` uses, because `CR_LOCATOR`'s third group is `(.*)$` and the 2024/2025 vintages put the title *on* the locator line. Reading it from the stripped copy empties 756 of the 1693 review-body findings.

## Linked Issues

None. Found while reading `harvest.mjs`'s two CodeRabbit readers side by side.

## Author checklist

- [x] I searched existing issues and PRs and confirmed this is not a duplicate.
- [x] I read every line of this diff and can explain why each change is necessary.
- [x] Changes follow the conventions in the touched packages; any deviations are explained in *Why* above.
- [x] If AI tools assisted with this PR, I noted where in *Notes for Reviewers* below.

## Verification

- [ ] verify:self — CI comment shows ✅
- [ ] verify:integration — CI comment shows ✅ (or explicit skip reason below)

Skip reason (if applicable): none — both lanes run. The two boxes are left unticked because CI has not reported on this branch yet; they are the reviewer's to confirm, not mine to pre-tick.

Locally, against a freshly measured baseline on `main` @ `d327ee4` in an identically-provisioned tree:

```
main   : 1696 + 55 = 1751 tests, 0 fail, 0 skipped
branch : 1701 + 55 = 1756 tests, 0 fail, 0 skipped   (+5)
npx eslint scripts: exit 0 on both
```

Two invocations, the shape the `agent:tests` lane uses since #774 — everything except `eval/run.test.mjs`, then that file alone. Both re-run from the committed tree rather than a working copy.

Every new test was mutation-tested: **7 mutations, 7 caught** — unbounding the title again (4 tests red), stripping only comments (3), stripping only fences (1), making the fence match greedy (2), giving the review-body path its own regex back (1), reading its title from the de-locatored copy (1), and a first-sentence fallback instead of `""` (2).

`harvest.mjs --audit` recovery counts are unchanged, per PR and in total, verified by running `auditCodeRabbit` on both trees with the endpoint payloads held fixed. A finding is counted on its header or its leading bold line, never on its title.

## Risk Assessment

- **User-facing risk:** none. `harvest.mjs` is an offline corpus tool. It is not on the review or gating path, records no check runs, and nothing it produces reaches a pull request.
- **Data/security risk:** this changes one field on CodeRabbit-derived findings, so any stored `summary` from a previous harvest differs from what a re-harvest now produces — and `summary` feeds `findingKey`, so 29 keys change. `detail` is byte-identical on both populations (1588 inline, 1693 review-body), no finding is gained or lost, and no finding gains an empty summary. `--audit` totals are unchanged.
- **Rollback plan:** revert. Three files, one of which is a task doc; no state, no migration.

## Notes for Reviewers

- **AI assistance:** written with Claude Code. It wrote the implementation, the tests and the task doc, ran the mutation testing, and produced every measurement quoted above by re-parsing this repository's own CodeRabbit output through both the current and the changed parser. I reviewed the diff and the numbers.
- **Follow-up work, deliberately not in this PR:** the same false premise in `CR_PROSE_END` leaves `detail` empty on 200 of those inline findings — findings that do have prose, which is why a CodeRabbit finding can end up compared on location alone. Fixing it means treating the prose region as the text *outside* `<details>` blocks. It is a strictly larger change and is described in the corrected comment above the constant so the next reader starts from the data rather than from the claim.
- `codeRabbitTitle` is exported so the two readers cannot drift apart again; the duplicated regex is exactly how they came to disagree.
- Task doc: `docs/tasks/active/20260812-coderabbit-title-excludes-code-todo.md`.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved CodeRabbit finding titles by excluding fenced code, HTML comments, and structured block content.
  - Preserved titles that appear after details sections, locator lines, or other formatted content.
  - Maintained existing finding counts and detailed descriptions.

- **Tests**
  - Added coverage for inline comments and review-body title extraction across supported formatting scenarios.

- **Documentation**
  - Documented title extraction behavior, edge cases, measured impact, and verification results.

### CODE REVIEW

File: scripts/agent/harvest.mjs [hunk 1]
```diff
@@ -488,6 +505,52 @@ const CR_PROSE_END = /^[ \t]*(?:<details>|```|<!--)/m;
...
+ * went from a real title to `""` — including all of the 12.6% described above
+ * `CR_PROSE_END`, whose titles sit after a `<details>` block. This rule changed 28,
+ * and every one of the 28 went from markup to a real title.
+ *
+ * Each span is matched LAZILY and removed on its own. A greedy `[\s\S]*` would run
+ * from the first fence to the last, and the ordinary shape of a CodeRabbit comment
+ * puts the title BETWEEN two of them — an analysis chain above, an AI-agents block
+ * below — so a greedy match eats exactly the string this function exists to find.
+ */
+const CR_NON_PROSE = /```[\s\S]*?```|<!--[\s\S]*?-->/g;
```
[the remark this answers is not in the document]
Comment (dlgpdmsly2) [reply scripts/agent/harvest.mjs]: Fixed in 8e9f46c — with one half of the suggestion adopted as correctness hardening and the other as widening, because they measure differently against this repository's own CodeRabbit output.

**Delimiter length: real shape, adopted.** A fence may legally contain a shorter one, and that is not hypothetical here — of 2061 CodeRabbit inline comments, **81 open a fence with four or more backticks and 35 of those wrap a three-backtick run**, which is exactly why an author reaches for four. Review bodies carry 59 more. The closer is now matched by backreference, so a shorter inner run cannot terminate a longer fence.

Being precise about the impact, though: **it changes no title in today's data — 0 of 1588 inline findings and 0 of 1693 review-body findings** differ between the fixed-length closer and the length-aware one. So this hardens a shape that exists rather than fixing a defect that has fired, and the comment above the constant says so rather than implying a repair.

**Tilde fences: not present, added anyway.** `~~~` appears in **0 of 2061** inline comments and **0 of 614** review bodies. Adopted as pure widening because the consequence if CodeRabbit ever emits one is markup becoming a finding's summary, which is what this function exists to prevent. The two delimiters cannot close each other.

**One thing I did not do, deliberately.** The rule stays anchored to line start, which leaves an inline ` ```**x**``` ` span unstripped. Widening to inline code spans is not a safe fix: a real title routinely quotes a symbol — this PR's own motivating example is *"Pin the markdown-it dependency or switch to the public `getRules()` API."* — and stripping spans before the bold search deletes the identifier out of the middle of the title being read. Measured, **39 comments carry a mid-line backtick run and none has a bolded run on that line**, so the guarded shape occurs zero times while the damage would apply to every title containing code. There is now a test that keeps a backticked identifier inside a title specifically to stop that gap being closed the wrong way; the wrong fix reddens 6 tests.

Tests added for both fence styles at both delimiter lengths, plus the unclosed-fence and inline-span cases: **1705 + 55 = 1760, 0 fail, 0 skipped** against a freshly measured 1696 + 55 = 1751 on `main`. Four new mutations, four caught — a fixed-length closer reddens 2, dropping tildes 1, an exactly-three opener 1, and stripping inline spans 6.

**Comment by dlgpdmsly2** [comment]
@claude review

**Review by hotsunchip — DISMISSED** [review]

File: scripts/agent/harvest.mjs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dlgpdmsly2) [reply scripts/agent/harvest.mjs]: Three findings; one was right and changes the shape of this PR, two are not valid as stated. Fixed in f26c5ed.

---

**1. Unfenced bold inside `🧩 Analysis chain` — CORRECT, and this was the important one.**

You're right, and the fixture hiding it is exactly the reason given: the trimmed content was the only content that could exhibit the failure. Skipping fenced code does not decide what a title is — the `💡 Result:` narrative is LLM prose, not code, and it uses bold freely.

Measured over all 2061 inline comments, the fence-only rule produced **16 titles that disagree with the finding's own prose**. Four are unambiguous web-answer fragments:

| comment | fence-only rule | the finding's actual title |
|---|---|---|
| 2876205560 | `Don't use` | Use `fileURLToPath` for filesystem paths derived from `import.meta.url`. |
| 2884648659 | `React 19` | Avoid side effects inside the `setDefinition` updater callback. |
| 3294479307 | `strings` | Restore `NODE_ENV` safely to avoid cross-test contamination. |
| 2902655936 | `` `DefaultLocale()` `` | Pin the collation used for group ordering. |

Same class of plausible-but-wrong summary as the `/node_modules/` glob this branch opened by fixing.

**The fix is that the title now reads the same prose region `codeRabbitDetail` does**, and that region learns what a `<details>` block is: a complete block is resolved before the boundary is searched for — machinery deleted, anything else unwrapped in place. That takes 16 → 5, and all 5 remaining are bodies whose prose is empty, so the comparison has no opinion there rather than a different one. The two readers now genuinely share one region, which is the only version of this that cannot drift: a `**` invisible to `detail` is invisible to the title.

Three sub-decisions, each forced by a measurement rather than chosen: **unwrap** non-machinery rather than delete it, because the 2025 `💡 Verification agent` vintage states its title and prose *inside* the block and deleting every block empties 11 of 1693 review-body findings; **dissolve innermost-first**, because a lazy match to the first closing tag cuts nested review bodies mid-structure; and **treat a closing tag as a boundary**, because a span inherits closers from elements that opened before it (735 of 1693 carried a `</details>` otherwise).

That also repairs `detail` as the same defect seen from the other side — `CR_PROSE_END` cut at the first `<details>`, so on those 200 findings the slice kept only the header line and the header-drop removed it. Empty details: **inline 200 → 5, review-body 3 → 1**, plus 154 pre-existing markup leaks cleared. Median length 331 → 345, max unchanged.

Since the machinery list is a denylist over a vocabulary CodeRabbit owns, it fails open — so `unrecognisedDetailsLabels` now reports what was unwrapped rather than recognised, and `--audit` prints it. It earned itself immediately: it surfaced five machinery labels my first list missed, including `🤖 Prompt for all review comments with AI agents`, a second phrasing of the block whose boilerplate this code exists to exclude.

---

**2. CRLF-intolerant closing fence — NOT VALID for JavaScript. Skipping, with a guard added.**

The reasoning is that with `m`, `$` matches only immediately before `\n`. In JavaScript `$` matches before any **LineTerminator**, and CR is one (ECMA-262 defines LineTerminator as LF, CR, LS, PS) — so on `` ``` `` + `\r\n` the position before `\r` is a valid `$`. That is a PCRE/Python behaviour, not a JS one.

Verified rather than argued: `/a$/m.test("a\r\nb")` is `true`; the real 9224-byte body of 3651715274 yields a byte-identical summary under LF and CRLF; and a 4-backtick-wrapping-3-backtick fence still strips correctly on CRLF. Separately, this repository's live bodies contain **0** `\r` across all 2061 CodeRabbit inline comments.

I've added a test asserting CRLF and LF agree for both `codeRabbitTitle` and `codeRabbitDetail`, and a mutation to a genuinely CR-intolerant pattern (`\n[ \t]*\1` instead of `^[ \t]*\1…$`) reddens 3 tests — so the claim is now answered by a guard rather than by this comment.

---

**3. Newly emits `summary: ""` — NOT VALID as stated. Skipping, with the reason.**

The change does not newly emit empty summaries. Across **1588 inline and 1693 review-body findings the count is unchanged from `main`: 0 and 2**, and both of the 2 predate this branch — they are `LGTM! …` notes with no bolded span anywhere, so `main` returns `""` for them too. The constructed fixture shows the shape is *reachable*, which is why it is pinned, but nothing in this repository's CodeRabbit output reaches it.

The hazard itself is real and I agree with it — which is why `codeRabbitTitle`'s docblock names it: `findingKey` is `` `${file}::${summary}` `` so empties collide in one file, and `findingSimilarity` scores two empty summaries at 1.00 (`rounds.mjs:525`), merging unrelated defects. But it is pre-existing, it lives in `finding-key.mjs` and `rounds.mjs`, and a fallback is the wrong fix here: a first-sentence or locator-derived title manufactures a summary CodeRabbit never wrote, and no consumer can then distinguish a manufactured one from a real one — the failure this PR's own argument rests on. The natural place to start is `validateFindingRecord`, which accepts `summary: ""` while explicitly refusing `run_id: ""` on the adjacent line. Left for its own change rather than smuggled in here.

---

**Verification:** **1714 + 55 = 1769**, 0 fail, 0 skipped, against a freshly measured **1696 + 55 = 1751** on `main` at `d327ee4`. eslint 0. **11 mutations, 11 caught**, including the two this round turned on. `--audit` recovery counts identical per PR and in total. 11 of 1588 inline summaries change against `main`, all from markup or a fragment to a real title, none to empty; review-body summaries unchanged; no finding gained or lost. On the frozen pilot corpus 1 of 30 summaries changes, all 3 empty details fill in, census still `30 / 0 / 0`.

One note on CI: `eval/run.test.mjs`'s `END TO END` cases are intermittently red locally on this machine — they reproduce on unmodified `main` at the same commit (3 failures across 4 runs there), and that file contains no reference to `harvest.mjs`, so they are not reachable from this change. Known shared-state flakiness around `os.tmpdir()`, from #682.

**Comment by dlgpdmsly2** [comment]
Addressed the remaining two findings from the panel's round in de4dde4, and correcting a number I got wrong in my previous reply.

**🔴 Correction first: I said "11 of 1588 inline summaries change against `main`". It is 38.** The 11 was the delta from the *previous commit on this branch*, quoted as if it were against `main`. Re-measuring caught it. The composition — all 38 go from markup or a fragment to a real title, and **none becomes empty**:

- **28** from the fenced-code rule: shell globs and transcripts (`/node_modules/`, `/dist/`, and 1.4 KB of `rg` output on comment 2911274299).
- **10** from sharing the machinery-stripped region: `Don't use`, `React 19`, `strings`, `` `DefaultLocale()` ``, plus five where a preliminary note inside an analysis chain gave way to the conclusion below it.

Review-body summaries change on 1 of 1693. Empty summaries are unchanged from `main` on both populations — 0 inline and 2 review-body — and both of the 2 predate this branch.

---

**(minor) A title could be spliced across a removed span — CORRECT, fixed.**

Confirmed on a constructed body: the function returned *"bold to start but never close it and then"*, text that appears in no line of the comment. That is worse than returning markup, because it reads like prose and nothing downstream can tell.

The bold match is now single-line. That makes the splice unrepresentable rather than unlikely, and it costs nothing measurable: **not one bolded span crosses a newline in 2061 inline comments**, so `[^\n]+?` and `/s` agree on every finding in this repository. It also removes the `/s` flag the finding named.

**(nit) A tilde run could close a backtick fence — CORRECT, fixed, and it needed a second half.**

`` \1[`~]* `` did accept ```` ```~~~ ````. Backticks and tildes are now separate alternatives with their own backreference.

Worth flagging that strictness *alone* makes the malformed case worse rather than better: the fence stops closing, so nothing is stripped, and a `**` sitting inside the code becomes the title — the outcome this whole change exists to prevent. Measured on `` ```~~~ ``: lax gives the real title by luck, strict-only gives `not a title` (markup), and strict-plus-consume-to-end gives `""`. So an unclosed fence now runs to the end, which is what CommonMark says. A title *above* such a fence is outside the span and still found.

Both halves are unobservable in today's data — **0 tilde fences, and no comment with an odd number of fence runs** — so this is a rule made to state something true rather than something lucky.

**That second fix changed a documented behaviour, and a test from the previous round went red for it.** It pinned *"an unclosed fence strips nothing"*, which is no longer true. The assertion and the docblock claim are corrected rather than deleted.

---

**Verification:** **1716 + 55 = 1771**, 0 fail, 0 skipped, against a freshly measured **1696 + 55 = 1751** on `main` at `d327ee4`. eslint 0. **14 mutations, 14 caught, 0 skipped** — including all five this round turned on. One mutation had been *skipping silently* on an earlier run because its target string had moved; the harness prints `TARGET NOT FOUND`, and that is worth grepping for rather than scrolling past. `--audit` recovery counts identical. Corpus: 1 of 30 summaries changes, all 3 empty details fill in, census still `30 / 0 / 0`.

**One note on the round itself:** 4 of the 7 findings under *Blocking* are `Review could not run — Claude API/quota error (429)` on Security, Design-fit, Test-adequacy and Blast-radius. So this diff has been read by **one lens of five**, and the three real findings all came from Correctness. Worth a re-run once quota resets before treating the review as coverage — three of the five findings it did produce were valid, and two of those were ones I had reasoned my way past rather than measured.

**Review by hotsunchip — APPROVED** [review]
